data-plane/receiver-loom/src/main/java/dev/knative/eventing/kafka/broker/receiverloom/LoomKafkaProducer.java

```diff
@@ -15,6 +15,145 @@
+import io.vertx.core.Vertx;
+import io.vertx.core.impl.ContextInternal;
+import io.vertx.core.impl.VertxInternal;
+import java.util.Objects;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.apache.kafka.clients.producer.Producer;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.clients.producer.RecordMetadata;
+import org.slf4j.Logger;
+
+public class LoomKafkaProducer<K, V> implements ReactiveKafkaProducer<K, V> {
+
+    Logger logger = org.slf4j.LoggerFactory.getLogger(LoomKafkaProducer.class);
+
+    private final Producer<K, V> producer;
+
+    private final BlockingQueue<RecordPromise> eventQueue;
+    private final AtomicBoolean isClosed;
+    private final ProducerTracer tracer;
+    private final VertxInternal vertx;
+    private final ContextInternal ctx;
+    private final Thread sendFromQueueThread;
+
+    public LoomKafkaProducer(Vertx v, Producer<K, V> producer) {
+        Objects.requireNonNull(v, "Vertx cannot be null");
+        this.producer = producer;
+        this.eventQueue = new LinkedBlockingQueue<>();
+        this.isClosed = new AtomicBoolean(false);
+        this.vertx = (VertxInternal) v;
+        this.ctx = vertx.getOrCreateContext();
+        ContextInternal ctxInt = ((ContextInternal) v.getOrCreateContext()).unwrap();
+        this.tracer = ProducerTracer.create(ctxInt.tracer());
+
+        sendFromQueueThread = Thread.ofVirtual().start(this::sendFromQueue);
+    }
+
+    @Override
+    public Future<RecordMetadata> send(ProducerRecord<K, V> record) {
+        Promise<RecordMetadata> promise = Promise.promise();
+        if (isClosed.get()) {
+            promise.fail("Producer is closed");
+        } else {
+            eventQueue.add(new RecordPromise(record, promise));
+        }
+        return promise.future();
+    }
+
+    private void sendFromQueue() {
+        while (!Thread.currentThread().isInterrupted()) {
+            try {
+                RecordPromise recordPromise = eventQueue.take();
+                ProducerTracer.StartedSpan startedSpan =
+                        this.tracer == null ? null : this.tracer.prepareSendMessage(ctx, recordPromise.getRecord());
+                try {
+                    var metadata = producer.send(recordPromise.getRecord());
+                    recordPromise.getPromise().complete(metadata.get());
+                    if (startedSpan != null) {
+                        startedSpan.finish(ctx);
+                    }
+                } catch (Exception e) {
+                    recordPromise.getPromise().fail(e);
+                    if (startedSpan != null) {
+                        startedSpan.fail(ctx, e);
+                    }
+                }
+            } catch (InterruptedException e) {
+                logger.debug("Interrupted while waiting for event queue to be populated.");
+            }
+        }
```

Comment:
```suggestion
        }
        
        logger.debug("Background thread finished.");
```

Reply:
I have just changed the `sendFromQueue` function to use the callback and it is stopping now.